event_model/__init__.py

```diff
@@ -120,6 +120,176 @@ def bulk_datum(self, doc):
         self.datum_page(bulk_datum_to_datum_page(doc))
 
 
+class Filler(DocumentRouter):
+    """Pass documents through, loading any externally-referenced data.
+
+    It is recommended to use the Filler as a context manager.  Because the
+    Filler manages caches of potentially expensive resources (e.g. large data
+    in memory) managing its lifecycle is important. If used as a context
+    manager, it will drop references to its caches upon exit from the
+    context. Unless the user holds additionally references to those caches,
+    they will be garbage collected.
+
+    But for some applications, such as taking multiple passes over the same
+    data, it may be useful to keep a longer-lived Filler instance and then
+    manually delete it when finished.
+
+    See Examples below.
+
+    Parameters
+    ----------
+    handler_registry : dict
+        Maps each 'spec' (a string identifying a given type or external
+        resource) to a handler class.
+
+        A 'handler class' may be any callable with the signature::
+
+            handler_class(resource_path, root, **resource_kwargs)
+
+        It is expected to return an object, a 'handler instance', which is also
+        callable and has the following signature::
+
+            handler_instance(**datum_kwargs)
+
+        As implied by the names, this is typically implemented using a class that
+        implements ``__init__`` and ``__call__``, with the respective signatures.
+        In general it may be any callable-that-returns-a-callable.
+    handler_cache : dict, optional
+        A cache of handler instances. If None, a dict is used.
+    datum_cache : dict, optional
+        A cache of Datum documents. If None, a dict is used.
+    retry_intervals : Iterable, optional
+        If data is not found on the first try, there may a race between the
+        I/O systems creating the external data and this stream of Documents
+        that reference it. If Filler encounters an ``IOError`` it will wait a
+        bit and retry. This list specifies how long to sleep (in seconds)
+        between subsequent attempts. An empty list means, "Try only once." If
+        None, a sensible default is used. That default should not be considered
+        stable; it may change at any time as the authors tune it.
+
+    Examples
+    --------
+    A Filler may be used as a context manager.
+
+    >>> with Filler(handler_registry) as filler:
+    ...     for name, doc in stream:
+    ...         filler(name, doc)
+
+    Or as a long-lived object.
+
+    >>> f = Filler(handler_registry)
+    >>> for name, doc in stream:
+    ...     filler(name, doc)
+    ...
+    >>> del filler  # Free up memory from potentially large caches.
+    """
+    def __init__(self, handler_registry, *,
+                 handler_cache=None, datum_cache=None, retry_intervals=None):
```

Comment:
Do we want to enable filling only certain fields?

Reply:
Good point! I think we do.